File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Fastuse.java
package me.zeroeightsix.kami.module.modules.combat;

import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.item.*;
import net.minecraft.network.play.client.CPacketPlayerDigging;
import net.minecraft.network.play.client.CPacketPlayerTryUseItem;
import net.minecraft.util.math.BlockPos;

/**
 * Created by S-B99 on 23/10/2019
 *
 * @author S-B99
 * Updated by S-B99 on 03/12/19
 * Updated by d1gress/Qther on 4/12/19
 * <p>
 * Bowspam code from https://github.com/seppukudevelopment/seppuku/blob/5586365/src/main/java/me/rigamortis/seppuku/impl/module/combat/FastBowModule.java
 */
@Module.Info(category = Module.Category.COMBAT, description = "Use items faster", name = "FastUse")
public class Fastuse extends Module {

    private static long time = 0;
    private Setting<Integer> delay = register(Settings.integerBuilder("Delay").withMinimum(0).withMaximum(20).withValue(0).build());
    private Setting<Boolean> all = register(Settings.b("All", false));
    private Setting<Boolean> bow = register(Settings.booleanBuilder().withName("Bow").withValue(true).withVisibility(v -> !all.getValue()).build());
    private Setting<Boolean> expBottles = register(Settings.booleanBuilder().withName("Exp Bottles").withValue(true).withVisibility(v -> !all.getValue()).build());
    private Setting<Boolean> endCrystals = register(Settings.booleanBuilder().withName("End Crystals").withValue(true).withVisibility(v -> !all.getValue()).build());
    private Setting<Boolean> fireworks = register(Settings.booleanBuilder().withName("Fireworks").withValue(false).withVisibility(v -> !all.getValue()).build());

    @Override
    public void onDisable() {
        mc.rightClickDelayTimer = 4;
    }

    @Override
    public void onUpdate() {
        if (mc.player == null) return;

        if (all.getValue() || bow.getValue() && mc.player.getHeldItemMainhand().getItem() instanceof ItemBow && mc.player.isHandActive() && mc.player.getItemInUseMaxCount() >= 3) {
            mc.player.connection.sendPacket(new CPacketPlayerDigging(CPacketPlayerDigging.Action.RELEASE_USE_ITEM, BlockPos.ORIGIN, mc.player.getHorizontalFacing()));
            mc.player.connection.sendPacket(new CPacketPlayerTryUseItem(mc.player.getActiveHand()));
            mc.player.stopActiveHand();
        }

        if (!(delay.getValue() <= 0)) {
            if (time <= 0) time = Math.round((2 * (Math.round((float) delay.getValue() / 2))));
            else {
                time--;
                mc.rightClickDelayTimer = 1;
                return;
            }
        }

        if (passItemCheck(mc.player.getHeldItemMainhand().getItem()) || passItemCheck(mc.player.getHeldItemOffhand().getItem())) {
            mc.rightClickDelayTimer = 0;
        }
    }

    private boolean passItemCheck(Item item) {
        if (all.getValue()) return true;
        if (expBottles.getValue() && item instanceof ItemExpBottle) return true;
        if (endCrystals.getValue() && item instanceof ItemEndCrystal) return true;
        return fireworks.getValue() && item instanceof ItemFirework;
    }
}

File: src/main/kotlin/org/kamiblue/client/module/modules/player/FastUse.kt
package org.kamiblue.client.module.modules.player

import net.minecraft.init.Items
import net.minecraft.item.*
import net.minecraft.network.play.client.CPacketPlayerDigging
import net.minecraft.network.play.client.CPacketPlayerTryUseItem
import net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.gameevent.TickEvent
import org.kamiblue.client.event.events.PacketEvent
import org.kamiblue.client.mixin.extension.rightClickDelayTimer
import org.kamiblue.client.module.Category
import org.kamiblue.client.module.Module
import org.kamiblue.client.util.threads.safeListener
import org.kamiblue.event.listener.listener

/**
 * Bowspam code from https://github.com/seppukudevelopment/seppuku/blob/5586365/src/main/java/me/rigamortis/seppuku/impl/module/combat/FastBowModule.java
 */
internal object FastUse : Module(
    name = "FastUse",
    category = Category.PLAYER,
    description = "Use items faster"
) {
    private val delay = setting("Delay", 0, 0..10, 1)
    private val blocks = setting("Blocks", false)
    private val allItems = setting("All Items", false)
    private val expBottles = setting("Exp Bottles", true, { !allItems.value })
    private val endCrystals = setting("End Crystals", true, { !allItems.value })
    private val fireworks = setting("Fireworks", false, { !allItems.value })
    private val bow = setting("Bow", true, { !allItems.value })
    private val chargeSetting = setting("Bow Charge", 3, 0..20, 1, { allItems.value || bow.value })
    private val chargeVariation = setting("Charge Variation", 5, 0..20, 1, { allItems.value || bow.value })

    private var lastUsedHand = EnumHand.MAIN_HAND
    private var randomVariation = 0
    private var tickCount = 0

    val bowCharge get() = if (isEnabled && (allItems.value || bow.value)) 72000.0 - (chargeSetting.value.toDouble() + chargeVariation.value / 2.0) else null

    init {
        safeListener<TickEvent.ClientTickEvent> {
            if (it.phase != TickEvent.Phase.END || player.isSpectator) return@safeListener

            if ((allItems.value || bow.value) && player.isHandActive && (player.activeItemStack.item == Items.BOW) && player.itemInUseMaxCount >= getBowCharge()) {
                randomVariation = 0
                connection.sendPacket(CPacketPlayerDigging(CPacketPlayerDigging.Action.RELEASE_USE_ITEM, BlockPos.ORIGIN, player.horizontalFacing))
                connection.sendPacket(CPacketPlayerTryUseItem(player.activeHand))
                player.stopActiveHand()
            }

            if (delay.value > 0) {
                if (tickCount <= 0) {
                    tickCount = delay.value
                } else {
                    tickCount--
                    return@safeListener
                }
            }

            if (passItemCheck(player.getHeldItem(lastUsedHand).item)) {
                mc.rightClickDelayTimer = 0
            }
        }

        listener<PacketEvent.PostSend> {
            if (it.packet is CPacketPlayerTryUseItem) lastUsedHand = it.packet.hand
            if (it.packet is CPacketPlayerTryUseItemOnBlock) lastUsedHand = it.packet.hand
        }
    }

    private fun getBowCharge(): Int {
        if (randomVariation == 0) {
            randomVariation = if (chargeVariation.value == 0) 0 else (0..chargeVariation.value).random()
        }
        return chargeSetting.value + randomVariation
    }

    private fun passItemCheck(item: Item): Boolean {
        return item !is ItemAir
            && (allItems.value && item !is ItemBlock
            || blocks.value && item is ItemBlock
            || expBottles.value && item is ItemExpBottle
            || endCrystals.value && item is ItemEndCrystal
            || fireworks.value && item is ItemFirework)
    }
}